all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+# External Client Configuration
+
+Notes for this proposal:
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+There are many ways to access Temporal programmatically including CLI, SDK, and other components that leverage those
+such as samples and Terraform providers.
+
+Temporal needs a way for users to provide client configuration without hardcoding in CLI arguments or SDK options. This
+will allow code to seamlessly switch between profiles such as between a dev-server and cloud. The same configuration
+must work across all SDKs/CLIs.
+
+### Why?
+
+What are the reasons for building this?
+
+* Users want to switch environments without changing their code, but today they cannot.
+* Our samples should be able to be run on multiple environments including cloud, but today they are localhost only.
+
+### Goals
+
+* Common, well-specified configuration format usable by Temporal tooling for client options
+* Implementation in all Temporal SDKs for reading and writing of this configuration
+* All Temporal CLI/SDKs updated to have easy client creation using this configuration
+* Common on-disk location in every platform for reading/writing this format
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
```

Comment:
> Why do we need this? 

Multiple profiles in the same configuration makes it easier to edit from CLI and have a single config to ship around with multiple profiles if you want.

> Can we not achieve the same thing in a simpler fashion by providing a way to set the configuration file path from the environment (for example)?

We offer this _too_, but the default is a single place.

> The only value I can think of for having them is in the CLI itself, where it's convenient to be able to refer to different environments by short names rather than full config file paths. But I'm not sure if that use case needs to bleed into the SDKs.

CLI and SDKs need to share the same approach here. A single config file that contains multiple profiles helps all clients IMO.

Reply:
>> CLI and SDKs need to share the same approach here.
> Why

Because the same configuration needs to work in all Temporal clients (SDKs and CLIs).

> The CLI has some use cases/scenarios that don't apply to the (other) SDKs, like being used for operational/admin tasks, possibly in multiple environments.

Sure, and Java has scenarios that don't apply to Python and so on. But this configuration should apply to both. I think AWS CLI vs SDK is a decent model here. I think consistency should be the default and we need a really good reason to be inconsistent.

> I am struggling to come up with a "multi-profile use case" that can't also easily be handled with multi-file. Can you provide an example?

Sure, technically everything you can do with a single file you can do with multi file, so you will struggle if you're trying to see a situation where we _must_ force multi file. Same goes with any config that has separate sections. For instance, you could say you're struggling to come up with a use case why the `tls` config option can't also easily be handled with multi-file. Why not have the TLS file separate, wouldn't this make downloading it and sharing it easier?

The reason we have nested keys in a configuration is for convenience of not having to ship around multiple files or deal with multiple files. You can load up an entire file at once and mutate multiple profiles. You can onboard a new dev by giving a single file with all profiles in it pointing to different clusters. It is easier to glance at all configuration in a file. The entire configuration for use in an SDK can be one set of bytes (some users may not use files, they may provide the bytes some other way). Etc. The same reasons you may want more than one profile in a file extend to wanting more than one key in a file.

I think justifying not using a separate file for separate keys (whether those keys are profiles or just nested options like `tls`) is more important. AWS and others put all profiles in a single file and I think we should too. When Temporal says "configuration" it should be a single thing, whether that's a proto object, a set of bytes in memory, a set of bytes on disk, etc. To have separate profiles be separate objects, separate sets of bytes, separate disk files, etc is a bit rough.

Reply:
> Regarding development contexts, consider an organization with a large monorepo containing many different projects that make use of Temporal. The different projects in the monorepo will want the ability to commit independent Temporal config files in their own subdirectories of the monorepo.

I do not agree with this statement. These configurations are not code configurations they are environment configurations. I can't imagine any reason to commit Temporal config files any more than AWS CLI/SDK config files. In fact, I believe it is important that these configuration values remain environment config only.

I would expect the dozens of configuration options that a user can set in the SDK to remain set in the SDK. Notice I didn't add all of the client options as configuration options. I only added ones that would be environment specific and have value not being in code or committed. Overrides and defaults can still be part of code. This configuration is only meant for non-committed data that changes from one environment to the next.

Reply:
Yes, I think this is the key thing we need to establish first. If this configuration file is just configuring the client connection, then it's just environment and wouldn't typically be committed.

The thing is that the proposal currently has the file name as `.../temporal.json` or `.../temporal.toml`. And that raises the following questions

1. Will Temporal want to support file-based configuration of things other than client connections?
2. If so, should the client connection be a stanza within a broader config file?

I understand that you'd prefer not to think about unknown things in the future, but if we're naming the file `temporal.{json,toml}` then unfortunately we don't have that option, since we'd be taking the name that a future general Temporal config file would use. (If we were to name it `temporal-client-config.{json,toml}`, then we could perhaps design things so that in the future it can also be supplied via a stanza within a more general `temporal.{json,toml}`).

Temporal aims to be a very general application development framework, and already has a fairly large product surface area. It would be a very bold statement at this point in history to say "Temporal won't ever want any file-based config beyond client connection configuration."

Concretely, what possibilities can we think of right now? Any individual point I list we'll probably be able to come up with criticisms of (i.e. why we might think it's not needed in file-based config), but the point is that there are already possibilities, and those will only increase as time passes; it seems likely to me that the name `temporal.{json,config}` will be wanted in the future for config that includes non-client-connection things.

- dev-server
  - start-up flags: always enable $experimental_feature via dynamic config flags
  - enable labs / experimental features in UI
  - file-based persistence config


- telemetry
  - configure collectors uniformly across all SDKs 
  - temporal developers may want to configure the same collector for server and SDK-side traces to obtain views of the entire workflow lifecycle
  - set SDK logging options so that they apply uniformly across CLI and any SDK


- CLI
  - dark mode vs light mode
  - enable/disable color/hyperlinks/decorations
  - hyperlink format (e.g. support creation of hyperlinks by the CLI that route to a different UI for viewing workflows) 


- Application gRPC logic?
  - I'm not sure about this -- it seems useful for development, but dangerous/inappropriate to risk having dev defaults applied in production.
  - Perhaps users would want to globally set certain "policy"-type options that are used in multiple gRPCs APIs.
    - E.g. workflow ID reuse/conflict policies? (I know I personally always want `TERMINATE_IF_RUNNING` when working on samples, but at the same time can't actually commit it to my samples).

Reply:
> Will Temporal want to support file-based configuration of things other than client connections?

Not to my knowledge. But maybe the better way to ask is "Will Temporal want to support file-based configuration of things other than environment-specific configuration?" because I think the environment-specific configuration is so minimal it can remain top-level. I think the answer to that question is "no".

> If so, should the client connection be a stanza within a broader config file?

I don't believe there is any concept of broader config for this use case. This is basically a file that can be used as an alternative to environment variables.

> It would be a very bold statement at this point in history to say "Temporal won't ever want any file-based config beyond client connection configuration."

I don't think it's too bold to say "Temporal won't ever want to reuse this file/env-based config for anything beyond environment configuration (which is only client so far)".

> dev-server

I can't think of anything about the dev server that belongs in this configuration (nor can I for any other CLI arg that is not already included in this doc). This is not a general purpose configuration, this is per-environment configuration. Basically anything you might use env vars for.

> telemetry

This I can see being environment configuration and I think it could live right alongside the current environment configuration. In fact, unlike the client options that are opt-in to taking from environment configuration, I can see making these implicitly loaded (there are caveats about reading from disk by default in libraries, but can be discussed).

> CLI

Not this kind of configuration IMO. Not only do we not want any CLI-specific configuration here IMO, but CLI args don't always qualify, including ones that affect output.

> Application gRPC logic?

I don't want to confuse that in this kind of configuration. We do have ways to set a lot of options, but those options are not part of what I would consider this configuration for these environment specific things. Sure one can say "but we want to have different conflict policy per environment", well sure anyone can technically say "we want to have different <option> per environment" and they are more than welcome to, but I don't think that goes in _this_ config.

Reply:
> This is basically a file that can be used as an alternative to environment variables.

> This is not a general purpose configuration, this is per-environment configuration.

> Basically anything you might use env vars for

> Not only do we not want any CLI-specific configuration here IMO

> I don't want to confuse that in this kind of configuration. We do have ways to set a lot of options, but those options are not part of what I would consider this configuration for these environment specific things.

> this file/env-based config

> I don't think that goes in this config.

In all of the above statements you're repeating an assertion -- but it's precisely that assertion that I'm questioning, so repeating the assertion doesn't add anything. The logic here (and I do think I made it clear above) is that

1. **Premise**: if Temporal will want to use file-based config for non-client-connection things in the future, then we cannot have connection-only config squatting on the name `temporal.json`

2. Therefore, we can only use that name if Temporal will not seek to use file-based config for non-client-connection-things in the future.

So we need to establish whether we'll want to use file-based config for things other than client connection. That's why I'm encouraging others to think about this / brainstorm what uses we might have for file-based config in the future. I thought about it a bit and have given some points above suggesting that the answer might be "yes, we'll want to have file-based config for things other than client connection".

Reply:
> if Temporal will want to use file-based config for non-client-connection things in the future, then we cannot have connection-only config squatting on the name temporal.json

I contend Temporal does not want to use file-based config for non-_environment_ configuration, and _environment_ configuration is currently all client config and in the future probably also all client config or almost surely the vast majority (maybe one or two runtime config options).

Reply:
Right, now we're making progress! So (correct me if I'm getting you wrong), you're basically using "environment config" with the following meaning:

> The SDKs and CLI all require certain resources to be specified (e.g. server address), together with things like encryption settings and credentials for those resources. The key point is that these _environment_ choices (kind of) don't change the behavior of the SDK/CLI -- they're about choosing from among a range of possible resource providers.

> This is in contrast to other configuration settings that change the behavior of the SDKs and CLIs.

And I agree, this is a distinction we should make.

I've given several reasons above to think that *non-environment* config is something we will want, whether sooner or later: I've given concrete suggestions for what those might be today, and I've also appealed to the fact that Temporal has a large product surface area, ambitions, and a lot of time in which it will exist in the future hence is likely to end up wanting this at some point.

So unless we can identify reasons why we in fact do not think we'll ever want non-environment config, the challenge here is to design file-based configuration in a way that supports the following requirements (but not limited to these):

1. It is possible to define non-environment config.
1. It is possible to add new top-level non-environment config sections in the future.
1. It is possible to define multiple alternative environment profiles.
1. It is possible to use env vars to supply values for all environment config values where an env var makes sense.

Reply:
> I've given concrete suggestions for what those might be today

But I think they should be rejected for config except for environment-specific suggestions, no matter how far in the future non-environmental config is suggested. It doesn't mean other avenues for non-environment things can't be explored, I just don't think we need to account for them as part of this config which should be environment config only.

I think we just have to decide if this config is environment-config only or not. I propose so, forever into the future.

Reply:
> I just don't think we need to account for them as part of this config which should be environment config only.

Right, I just don't think we can use the name `temporal.{json,toml}` if it's never going to have general config in it. So maybe

- `temporal-profiles.{json,toml}`
- `temporal-client-config.{json,toml}`


Alternatively, perhaps we should be less strict and call it `temporal.toml` and allow that it might acquire non-environment config in the future, as other top-level keys that are siblings of the `profiles` key.

Reply:
To be clear: I don't really like the idea of multiple files. I think users will think of it all as "config", and will be annoyed that Temporal has "multiple different config files", and so it will make the product seem more complicated. So, even though we might like to avoid a difficult design process at this stage, that might not be the right decision for the product.

Reply:
We can change the default filename, no problem. I have no strong preference on which alternative there is chosen. I suspect people won't really use it directly anyways very much because if they want to hand-edit a file they might put it somewhere more reasonable and provide that filename when loading or as an env var. Or they'll use the CLI. Similar to AWS config.

I would be concerned however if we need to change the env var names or property names or the CLI commands or anything. I do think from a user POV we still need to treat this as "the config" and not inject client in those places.